CONTRIBUTING.md

```diff
@@ -24,19 +24,79 @@ Please follow these steps when making pull requests (PRs):
-        may have been updated after starting your PR.
+        may have been updated *after* starting your PR.
     Your PR **MUST** be updated to reflect changes to `master` in order to be merged.
-    Use the following procedure for updating your branch:
+    Use the following procedure for updating your branch and are ready to commit your changes:
 
     <!--- language: lang-shell -->
 
-        # While on your PR branch <issue-branch>
+        ## While on your PR branch <issue-branch> hosted at <your-remote> repository:
         git commit
         git checkout master
-        git pull origin --rebase
+        git pull <illixr-remote> --rebase                           ## Line A
         git checkout <issue-branch>
-        git rebase master
-        git push origin <issue-branch> --force-with-lease
+        git rebase master                                           ## Line B
+        git push <your-remote> <issue-branch> --force-with-lease    ## Line C
+
+    If you want to update your branch while you are still making changes to your branch:
+
+    <!--- language: lang-shell -->
+
+        ## While on your PR branch <issue-branch> hosted at <your-remote> repository:
+        git stash
+        git checkout master
+        git pull <illixr-remote> --rebase                           ## Line A
+        git checkout <issue-branch>
+        git rebase master                                           ## Line B
+        git stash apply <stash-number>
+        ## Finish your changes
+        git commit
+        git push <your-remote> <issue-branch> --force-with-lease    ## Line C
+
+    For ILLIXR team members:
+
+    -   In the examples above, `<illixr-remote>` and `<your-remote>` are the same.
+
+    -   When collaborating on branches in our repository, `Line A` may pull in changes that overwrite
+            the git commit history when performing `Line B`.
+        Subsequently, performing `Line C` will rewrite the history in the public branch.
+        To preserve branch commit histories in the case that a rollback is needed, we will employ
+            a checkpointing process for force updated branches.
+        This process will be manually performed, but may be automated in the future.
+
+        If `Line A` shows an update to master, the following example illustates your local repository
+            just after performing `Line A`:
+
+            A -- B -- C -- P -- Q -- R                  ## master
+                       \
+                        D -- E -- F                     ## issue-123-fixing-bug
+
+        To checkpoint the `issue-123-fixing-bug` while it is checked out:
+
+        <!--- language: lang-shell -->
+
+            git branch issue-123.0-fixing-bug                                   ## Make alias for old issue-123-fixing-bug
+            git checkout -b issue-123.1-fixing-bug                              ## Make new branch to rebase with master
```

Comment:
Shouldn't `issue-123.1-fixing-bug` be created after the rebase?

Reply:
The branch is created and checked out here as an alias of `issue-123.0-fixing-bug`. The following rebase will then modify `issue-123.1-fixing-bug` in-place, assuming the rebase completes successfully (including handling conflicts). After this process, `issue-123.0-fixing-bug` remains at the old history. If we rebase first, we need to make sure that the resulting branch is at a new, unlabeled detach head for which we then apply an alias/rename. We also need to make sure to preserve the old history.